flocker/node/agents/blockdevice.py

```diff
@@ -600,22 +608,33 @@ class DetachVolume(PRecord):
     """
     Detach a volume from the node it is currently attached to.
 
-    :ivar BlockDeviceVolume volume: The volume to destroy.
+    :ivar UUID dataset_id: The unique identifier of the dataset associated with
+        the volume to detach.
     """
-    volume = _volume_field()
+    dataset_id = field(type=UUID, mandatory=True)
 
     @property
     def eliot_action(self):
-        return DETACH_VOLUME(_logger, volume=self.volume)
+        return DETACH_VOLUME(_logger, dataset_id=self.dataset_id)
 
     def run(self, deployer):
         """
         Use the deployer's ``IBlockDeviceAPI`` to detach the volume.
         """
-        # Make this asynchronous after FLOC-1549, probably as part of
-        # FLOC-1593.
-        deployer.block_device_api.detach_volume(self.volume.blockdevice_id)
-        return succeed(None)
+        api = deployer.async_block_device_api
+        listing = api.list_volumes()
+        listing.addCallback(
+            _blockdevice_volume_from_datasetid, self.dataset_id
+        )
+
+        def found(volume):
+            if volume is None:
+                # It was not actually found.
+                raise DatasetWithoutVolume(dataset_id=self.dataset_id)
+            DETACH_VOLUME_DETAILS(volume=volume).write(_logger)
```

Comment:
:question: Use of `Logger` is no longer necessary (but you'd perhaps have to change tests correspondingly).